### PR TITLE
Add opcodes for EC operations

### PR DESCRIPTION
**Pull request by hkalodner**
- ECADD, ECMUL, ECPAIRING added to cpp avm
- ECADD, ECMUL, ECPAIRING added to ethbrige
- Add ECADD, ECMUL, ECPAIRING benchmarking and one step proof test
- ECADD gas cost is 3500
- ECMUL gas cost is 82000
- ECPAIRING gas cost is 1000 + min(k, 30) * 500000

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Clones added
============
- packages/arb-avm-cpp/tests/opcodes.cpp  2
- packages/arb-avm-cpp/tests/ecops.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/ecops.cpp  2
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/opcodes.cpp  2
- packages/arb-avm-cpp/tests/ecops.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/ecops.cpp  2
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/avm/src/machinestate/machineoperation.cpp  2
- packages/arb-avm-cpp/tests/opcodes.cpp  2
- packages/arb-avm-cpp/tests/ecops.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Review by edfelten — CHANGES_REQUESTED** [review]
Looks good, just two requests.

File: packages/arb-avm-cpp/avm/include/avm/machinestate/machineoperation.hpp [hunk 1]
```diff
@@ -23,6 +23,7 @@ struct MachineState;
 class Tuple;
 
 const int send_size_limit = 10000;
+constexpr int max_ec_pairing_points = 100;
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/avm/include/avm/machinestate/machineoperation.hpp]: Fixed

File: packages/arb-avm-cpp/avm/src/machinestate/machineoperation.cpp [hunk 2]
```diff
@@ -722,6 +733,113 @@ void ec_recover(MachineState& m) {
...
+        m.state = Status::Error;
+        return;
+    }
+
+    m.stack[0] = ret.get<bool>() ? 1 : 0;
+    ++m.pc;
+}
+
+uint64_t ec_pairing_variable_gas_cost(const MachineState& m) {
+    uint64_t gas_cost = 0;
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/avm/src/machinestate/machineoperation.cpp:822]: Fixed

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/ecops.cpp  2
- packages/arb-avm-cpp/avm/src/machinestate/machineoperation.cpp  2
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6198618&bid=20219723)

**Review by edfelten — APPROVED** [review]
Looks good.

**Pull request merged**